### PR TITLE
Multiple changes in AODProducerWorkflowSpec

### PR DESCRIPTION
**Pull request by shahor02**
### Multiple changes in AODProducerWorkflowSpec
    
*   Store ambiguous tracks only once, filling their BCslice corresponding to the time bracket as used in track-vertex matching.
*  Store BC slice for tracks w/o vertices
*  The time of ambiguous and orphan tracks is stored in ns wrt the globalBC corresponding to   bcSlice[0].
*  Add time info for MFT tracks
* Store in flags (at the moment for the barrel tracks only, since Fwd and MFT tracks have no  flags) the bit for the track being orphan, as well as the track time error being gaussian or  half-length of the range.
* Store in V0s and Cascades the collisonID.
* Add separate truncation data member for tracks time error.
* To guarantee required precision (set to 16ps) of the TOF tracks time the max distance to eventual reference globalBC is automatically estimated and if needed, dummy globalBCs are added to BCs table.

*   Add MFT track time and resolution to DataModel

*   Add flag for error being half-interval (non-gaussian)

*   Fix decomposition of TRD tracks indices

@jgrosseo The MFT tracks had no time info so I've added it to the model: https://github.com/shahor02/AliceO2/blob/384212ca1185de45b1885b50ab764677bb584890/Framework/Core/include/Framework/AnalysisDataModel.h#L463 but for some reason calling the cursor with time and error added does not work: if I uncoment https://github.com/shahor02/AliceO2/blob/384212ca1185de45b1885b50ab764677bb584890/Detectors/AOD/src/AODProducerWorkflowSpec.cxx#L277-L279 compilation fails with the complaint the too many arguments supplied. Any idea what could be the reason?

I see that for the Fwd and MFT tracks the precision truncation is not applied, was this done on purpose.

Currently, the flags exist only for barrel tracks (in Extra...), shall we add them also for the Fwd and MFT tracks?

### CODE REVIEW

**Comment by shahor02** [comment]
Removing WIP to run tests, do not merge even if green.

**Comment by ktf** [comment]
The error is genuine.

**Comment by jgrosseo** [comment]
"@jgrosseo The MFT tracks had no time info so I've added it to the model: https://github.com/shahor02/AliceO2/blob/384212ca1185de45b1885b50ab764677bb584890/Framework/Core/include/Framework/AnalysisDataModel.h#L463 but for some reason calling the cursor with time and error added does not work: if I uncoment https://github.com/shahor02/AliceO2/blob/384212ca1185de45b1885b50ab764677bb584890/Detectors/AOD/src/AODProducerWorkflowSpec.cxx#L277-L279 compilation fails with the complaint the too many arguments supplied. Any idea what could be the reason?"

-> Tables are redefined in the AODProducerWorkflowSpec.h (not sure if that could be done better). You need to add the columns also there.

**Comment by jgrosseo** [comment]
"I see that for the Fwd and MFT tracks the precision truncation is not applied, was this done on purpose." 
--> No, we should do it.

**Comment by jgrosseo** [comment]
- Thanks for adding the TrackTime to MFT. Do we need a resolution actually given that it is only one detector contributing?
- A FwdTrack (which has already TrackTime) references an MFTTrack and therefore such a track has two TrackTimes defined. It is not a problem, I mention it for completeness.
- I agree that we should add Flags to MFT and FwdTrack. For MFT it can be done like this, for FwdTrack it invalidates the converted data, so we need to apply there already now the schema evolution...

File: Detectors/AOD/src/AODProducerWorkflowSpec.cxx
Comment (jgrosseo) [inline Detectors/AOD/src/AODProducerWorkflowSpec.cxx]: `v0.getVertexID()` is the entry number in the collision table?

File: Detectors/AOD/src/AODProducerWorkflowSpec.cxx
Comment (jgrosseo) [inline Detectors/AOD/src/AODProducerWorkflowSpec.cxx:1051]: Why do we have different defaults here than above? (`-999`)

File: Detectors/AOD/src/AODProducerWorkflowSpec.cxx
Comment (shahor02) [reply Detectors/AOD/src/AODProducerWorkflowSpec.cxx]: you are right, for multi-TF processing this will create problems, will use mapping from vertexID to collisionID. But them I am not sure if the old code is correct: the reconstructed track indices are mapped to stored AOD index as https://github.com/AliceO2Group/AliceO2/blob/30327663640c088193254dd45d3e3a6f92040a1f/Detectors/AOD/src/AODProducerWorkflowSpec.cxx#L364-L365 (used to indicate V0's prong's indices). 
But they are reset after processing each TF:   https://github.com/AliceO2Group/AliceO2/blob/30327663640c088193254dd45d3e3a6f92040a1f/Detectors/AOD/src/AODProducerWorkflowSpec.cxx#L1475-L1476
It is OK to reset the map, since the reco indices are specific for each TF, but resetting the `mTableTrID` will break the correspondence to AOD track index (which is accumulated from one TF to another), right?

File: Detectors/AOD/src/AODProducerWorkflowSpec.cxx
Comment (shahor02) [reply Detectors/AOD/src/AODProducerWorkflowSpec.cxx:1051]: sorry, did not understand the question: the lines you refer to are compression masks (when no compression is requested, what they have to do with -999?)

**Comment by rpezzi** [comment]
> Thanks for adding the TrackTime to MFT. Do we need a resolution actually given that it is only one detector contributing?

MFT track time resolution depends on the readout configuration defined at the start of each run. There should be an optimal way to store this and propagate to AOD. Notice that there are also corner cases at the interface of ROFs where tracks are reconstructed only after merging clusters in two consecutive read out frames - this case has not yet been addressed in the MFT reconstruction but could be handled by a flag.

We have not foreseen MFT flags yet, but it may come handy in other cases as well. What should be a good place for flagging MFT tracks? MFTTrackExtra or add it to the main MFTTrack table?

Thanks a lot.

File: Detectors/AOD/src/AODProducerWorkflowSpec.cxx
Comment (jgrosseo) [reply Detectors/AOD/src/AODProducerWorkflowSpec.cxx:1051]: Sorry, forget about it.

**Comment by shahor02** [comment]
>> >> "I see that for the Fwd and MFT tracks the precision truncation is not applied, was this done on purpose."

>> --> No, we should do it.

OK, will add this

>> Do we need a resolution actually given that it is only one detector contributing?

ITS and MFT tracks time uncertainties depend on used strobe lengths which are the parameters of the run. Currently, they are constant within the run (so will be compressed out) but this may change when the tracking will properly account the tracks on the edge of the strobe (with clusters found in 2 strobes the time resolution will be on the level of ~100 ns).

>>  A FwdTrack (which has already TrackTime) references an MFTTrack and therefore such a track has two TrackTimes defined. It is not a problem, I mention it for completeness.

1st of all, the time of FwdTrack is more precise and different from the corresponding MFT track. And only a tiny fraction of MFT tracks have the associated FwdTrack. But in general, the way the MFT and Fwd tracks are related to each other breaks the "minimum redundancy" approach used for the barrel: e.g. ITS tracks participating in the global tracks are not counted as independent tracks (not stored in the AOD, instead, some of their features are stored in the Extra of the global track). 
In principle, the should have been done for the Fwd and MFT track on the level of RecoContainer selection: flag also MFT tracks as used in https://github.com/AliceO2Group/AliceO2/blob/30327663640c088193254dd45d3e3a6f92040a1f/DataFormats/Detectors/GlobalTracking/include/DataFormatsGlobalTracking/RecoContainerCreateTracksVariadic.h#L284
Don't know if this was not done on purpose (given the tiny amount of the ForwardTracks). Pinging to @rpezzy and @aphecetche. 

>>   I agree that we should add Flags to MFT and FwdTrack. For MFT it can be done like this, for FwdTrack it invalidates the converted data, so we need to apply there already now the schema evolution...

Will you add it? I am not feeling confident and modifications in the data model.

File: Detectors/AOD/src/AODProducerWorkflowSpec.cxx
Comment (jgrosseo) [inline Detectors/AOD/src/AODProducerWorkflowSpec.cxx:1840]: So you now add the dummy one closest to the tof time, right?
An alternative would be to use a bc number which actually had a valid bunch crossing. So all BCs would sit where we were sure true bunch crossings have taken place (even if it did not trigger any data taking).

File: Framework/Core/include/Framework/DataTypes.h
Comment (jgrosseo) [inline Framework/Core/include/Framework/DataTypes.h:38]: As you are on it, could you remove `ITSStandaloneTrack`  in line 32. I think it is not needed anymore.

File: Detectors/AOD/src/AODProducerWorkflowSpec.cxx
Comment (shahor02) [reply Detectors/AOD/src/AODProducerWorkflowSpec.cxx:1840]: This will not work for the cosmics. Also, for the beam with a few BCs only it is not guaranteed that you can find a BC with the `maxGapBC`. In any case, this fake BC is merely a reference.

File: Detectors/AOD/src/AODProducerWorkflowSpec.cxx
Comment (jgrosseo) [reply Detectors/AOD/src/AODProducerWorkflowSpec.cxx]: Yes, it does not look sane.
Better in any case would be `mGIDToTableID.emplace(trackIndex, tracksCursor.lastIndex()); `

File: Framework/Core/include/Framework/DataTypes.h
Comment (shahor02) [reply Framework/Core/include/Framework/DataTypes.h:38]: ok, done

File: Detectors/AOD/src/AODProducerWorkflowSpec.cxx
Comment (jgrosseo) [reply Detectors/AOD/src/AODProducerWorkflowSpec.cxx]: @aalkin @ktf Is `lastIndex` not defined for the cursor? We use it in O2Physics task on something defined by `Produces`. How can one have the last insert id here?

**Comment by jgrosseo** [comment]
> > I agree that we should add Flags to MFT and FwdTrack. For MFT it can be done like this, for FwdTrack it invalidates the 
> > converted data, so we need to apply there already now the schema evolution...

> Will you add it? I am not feeling confident and modifications in the data model.

Yes, I put it on my list.

File: Detectors/AOD/src/AODProducerWorkflowSpec.cxx
Comment (jgrosseo) [reply Detectors/AOD/src/AODProducerWorkflowSpec.cxx:1840]: I thought 1 BC per turn gives us already 4.5 ps precision and here you aim at 16ps?
Yes, cosmics needs to be treated differently.

File: Detectors/AOD/src/AODProducerWorkflowSpec.cxx
Comment (shahor02) [reply Detectors/AOD/src/AODProducerWorkflowSpec.cxx]: ok, I'll simply avoid resetting the counters like mTableTrID

File: Detectors/AOD/src/AODProducerWorkflowSpec.cxx
Comment (shahor02) [reply Detectors/AOD/src/AODProducerWorkflowSpec.cxx:1840]: Wth the maxGap=orbit and for the non-truncated float the precision in ps is indeed 
```
88924.e3/(1<<24)
(double) 5.3002834
```
but with default truncation of 8 lowest bits it is not enough:
```
88924.e3/(1<<16)
(double) 1356.8726
```
So, I am calculating the maxGap depending on the requested precision and adding fake BC to respect it.

**Comment by rpezzi** [comment]
> In principle, the should have been done for the Fwd and MFT track on the level of RecoContainer selection: flag also MFT tracks as used in
> 
> https://github.com/AliceO2Group/AliceO2/blob/30327663640c088193254dd45d3e3a6f92040a1f/DataFormats/Detectors/GlobalTracking/include/DataFormatsGlobalTracking/RecoContainerCreateTracksVariadic.h#L284
> Don't know if this was not done on purpose (given the tiny amount of the ForwardTracks). 
Hi Ruben, this was not done on purpose. We do not intend to exclude used MFT or MCH tracks used in global forward tracks.

**Comment by shahor02** [comment]
Hi @rpezzi 

>> Hi Ruben, this was not done on purpose. We do not intend to exclude used MFT or MCH tracks used in global forward tracks.

OK, so actually it was done on purpose: when filling MFT-MCH tracks you don't flag its constituent MFT and MCH as used, so they are provided to the data requestor on top of the global tracks.

**Comment by shahor02** [comment]
@jgrosseo Shall I merge it or it will affect the O2Physics? I still did not understand if recovering multiple vertices from the BCSlice is already supported.

**Comment by jgrosseo** [comment]
I'd merge it. Not everything is supported but this should not cause regressions and is an important step to the final data model.

**Review by jgrosseo — APPROVED** [review]

**Pull request merged**
